### PR TITLE
Change build order for post_config and post_rules

### PR DESCRIPTION
**Pull request by drashna**
## Description

Fully extends rules+config processing to userspace and enforce generic to specific flow.

Redo of #14852 because apparently force pushing prevents re-opening.

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).
  - Moderately tested on local boards. Doesn't appear to be broken.

### CODE REVIEW

**Comment by drashna** [comment]
@filterpaper this should work better for you, if using the post_rules.mk. 

Also, this definitely needs testing.

**Comment by spidey3** [comment]
I have tested this with `layouts/community/75_ansi/spidey3`, and it does work — but I did have to split `users/spidey3/rules.mk` — moving some of the directives to `users/spidey3/post_rules.mk`.

I suppose that is OK, given that this is targeted for the breaking change cycle...

File: docs/hardware_keyboard_guidelines.md [hunk 1]
```diff
@@ -88,13 +88,16 @@ The `config.h` files can also be placed in sub-folders, and the order in which t
...
           * [`.build/objs_<keyboard>/src/info_config.h`](data_driven_config.md#add-code-to-generate-it) see [Data Driven Configuration](data_driven_config.md)
-          * `users/a_user_folder/config.h`
-          * `keyboards/top_folder/keymaps/a_keymap/config.h`
-        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_config.h`
-      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_config.h`
-    * `keyboards/top_folder/sub_1/sub_2/post_config.h`
-  * `keyboards/top_folder/sub_1/post_config.h`
-* `keyboards/top_folder/post_config.h`
+            * `keyboards/top_folder/keymaps/a_keymap/config.h`
+              * `users/a_user_folder/config.h`
```
Comment (spidey3) [inline docs/hardware_keyboard_guidelines.md:92]: Should document loading of community layout `config.h` here...

File: docs/hardware_keyboard_guidelines.md [hunk 2]
```diff
@@ -144,12 +147,12 @@ The `rules.mk` file can also be placed in a sub-folder, and its reading order is
       * `keyboards/top_folder/sub_1/sub_2/sub_3/rules.mk`
         * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/rules.mk`
           * `keyboards/top_folder/keymaps/a_keymap/rules.mk`
-          * `users/a_user_folder/rules.mk`
-        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
+            * `users/a_user_folder/rules.mk`
```
Comment (spidey3) [inline docs/hardware_keyboard_guidelines.md:150]: Should document loading of community layout `rules.mk` here...

**Review by spidey3 — CHANGES_REQUESTED** [review]
Should document how community layout factors into loading order of `config.h` and `rules.mk`.

File: docs/hardware_keyboard_guidelines.md (hunk 1 above)
Comment (drashna) [reply docs/hardware_keyboard_guidelines.md]: Done, in both spots, as well as userspace.

**Review by spidey3 — APPROVED** [review]

**Comment by drashna** [comment]
> I have tested this with `layouts/community/75_ansi/spidey3`, and it does work — but I did have to split `users/spidey3/rules.mk` — moving some of the directives to `users/spidey3/post_rules.mk`.
> 
> I suppose that is OK, given that this is targeted for the breaking change cycle...

Well, that's why it's a "breaking" change. 🤣 

But yeah, this checks keyboard -> userspace -> keymap, so this will break a lot of user spaces, I think.  But this allows for better (and proper) handling, long term.

**Comment by drashna** [comment]
This breaks `USER_NAME = (name)` in keymap's rules.mk.  

Not sure there is a way to properly support it without some overly complicated makefile BS

**Pull request closed without merging**